File: php/callback.php
<?php

class CallBack
{
  private static $api_key = 'items';
  var $_params = array();
  
  function __construct($params)
  {
    $this->_params = $params;
  }
  
  function __get ($k) {
		if (isset ($this->_params[$k])) {
			return $this->_params[$k];
		}
		return $this->{$k};
	}
	  
  static function set_api_key($key){
    self::$api_key = $key;
  }
	
	function valid(){
    return($this->hmac() == $this->_params['signature']);
	}
	
	private function hmac(){
	  $data = $this->_params['timestamp'].
	          $this->_params['id'].
	          $this->_params['state'].
	          $this->_params['data']. 
	          $this->_params['src'];
	  return(hash_hmac('SHA1', $data, self::$api_key, false));
	}
}

?>

File: ruby/lib/socialmod/callback.rb
module Socialmod
  class Callback
    attr_reader :params,
                :id,
                :timestamp,
                :custom_id,
                :state,
                :signature,
                :time,
                :data,
                :src
                
    def initialize(params)
      @params    = params
      @id        = params['id']
      @timestamp = params['timestamp']
      @custom_id = params['custom_id']
      @state     = params['state']
      @signature = params['signature']
      @time      = Time.at(@timestamp.to_i)
      @data      = params['data']
      @src       = params['src']
    end
    
    def valid?
      hmac == @signature
    end
    
    def passed?
      state == 'passed'
    end
  
    def failed?
      state == 'failed'
    end
    
    def deferred?
      state == 'deferred'
    end
    
    def item
      id && Socialmod::Item.find(id)
    end
    
    protected
      def api_key
        Socialmod::Base.api_key || raise('Must provide API key')
      end
      
      def hmac
        str = [@timestamp, @id, @state, @data, @src].join('')
        @hmac_digest ||= OpenSSL::Digest::Digest.new('SHA1')
        OpenSSL::HMAC.hexdigest(@hmac_digest, api_key, str)
      end
  end
end